modules/ctransformers_model.py

```diff
@@ -9,29 +12,57 @@ class CtransformersModel:
     def __init__(self):
         pass
 
-    @classmethod
     def from_pretrained(self, path):
-        result = self()
 
         config = AutoConfig.from_pretrained(
             str(path),
             threads=shared.args.threads,
             gpu_layers=shared.args.n_gpu_layers,
             batch_size=shared.args.n_batch,
+            context_length=shared.args.n_ctx,
             stream=True
         )
 
+        model_name_dir = self.model_dir(path)
         self.model = AutoModelForCausalLM.from_pretrained(
-            str(result.model_dir(path) if result.model_type_is_auto() else path),
-            model_type=(None if result.model_type_is_auto() else shared.args.model_type),
+            str(model_name_dir if self.model_type_is_auto() else path),
+            model_type=(self.model_try_infer_type(model_name_dir, shared.args.model_type)),
             config=config
         )
 
         logger.info(f'Using ctransformers model_type: {self.model.model_type} for {self.model.model_path}')
-        return result, result
+        return self, self
 
     def model_type_is_auto(self):
-        return shared.args.model_type == "Auto" or shared.args.model_type == "None"
+        return shared.args.model_type is None or shared.args.model_type == "Auto" or shared.args.model_type == "None"
+
+    # llama is more of a fallback option
+    match_table = OrderedDict({
+        'falcon': re.compile('.*falcon.*ggml.*\\.bin'),
+        'mpt': re.compile('.*mpt.*ggml.*\\.bin'),
+        'starcoder': re.compile('.*star(chat|coder).*ggml.*\\.bin'),
+        'llama': re.compile('.*ggml.*\\.bin')
+    })
+
+    def model_try_infer_type(self, model_name_dir, type):
+        # User has explicitly selected the model type
+        if not self.model_type_is_auto():
+            return shared.args.model_type
+        # User has not selected but a config.json exists, let ctransformers handle it
+        if Path(f'{model_name_dir}/config.json').exists():
+            return None
+
+        # Try guessing
+        file_list = list(model_name_dir.glob('*.bin'))
+        if len(file_list) > 0 and file_list[0] is not None:
+            for type_match in self.match_table.keys():
+                pattern = self.match_table.get(type_match)
+                if(pattern.match(str(file_list[0].name))):
+                    logger.info(f'Guessing {file_list[0].name} is {type_match}')
+                    return type_match
```

Comment:
This part should be removed, see the comment above on `models/config.yaml`.

Reply:
Done.